### PR TITLE
[Console][DX] Fixed ambiguous error message when using a duplicate option shortcut

### PR DESCRIPTION
**Pull request by peterrehm**
| Q | A |
| --- | --- |
| Branch? | 2.3 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18856 |
| License | MIT |
| Doc PR | - |

I assume this should be merged into 2.3 as per @stof's comment.

There is a race condition when you run a command which has a duplicate option shortcut. Simply changing the order so that Options are merged before the Arguments solves that race condition.

``` php
$this->setName('my:super:command')
->setAliases(['my:super:commandalias'])
->setDescription('Performs some irrelevant work.')
->addOption('survey', 'e', InputOption::VALUE_REQUIRED, 'My option with a shortcut.')
```

Gives the error message:

```
  [Symfony\Component\Console\Exception\LogicException]
  An argument with name "command" already exists.
```

This happens as the first time the definition is merged happens here:

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Application.php#L820

As this throws an error here:

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Command/Command.php#L309

The commans are merged but not the options. 

Merging it then again when the command is run

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Command/Command.php#L217

throws an error due to the duplicate argument as the arguments already have been merged. This time the error message is not surpressed and will confuse the user.

Changing the order should fix the issue for duplicate arguments as well as for duplicate options.

### CODE REVIEW

**Comment by stof** [comment]
Please add a test covering this to prevent regressions

**Comment by stof** [comment]
btw, it may even go into 2.3 (if it is merged in the next few days)

**Comment by peterrehm** [comment]
@stof Test just has been added.

File: src/Symfony/Component/Console/Tests/ApplicationTest.php [hunk 1]
```diff
@@ -736,6 +736,33 @@ public function testRunReturnsExitCodeOneForExceptionCodeZero()
...
+            ->setDefinition(array(new InputOption('survey', 'e', InputOption::VALUE_REQUIRED, 'My option with a shortcut.')))
+            ->setCode(function (InputInterface $input, OutputInterface $output) {})
+        ;
+
+        $input = new ArrayInput(array('command' => 'foo'));
+        $output = new NullOutput();
+
+        $application->run($input, $output);
+    }
+
```
Comment (xabbuh) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: Please remove the trailing spaces here.

**Comment by xabbuh** [comment]
:+1: apart from my minor comment

Status: Reviewed

File: src/Symfony/Component/Console/Tests/ApplicationTest.php (hunk 1 above)
Comment (peterrehm) [inline src/Symfony/Component/Console/Tests/ApplicationTest.php]: Updated.

**Comment by fabpot** [comment]
Thank you @peterrehm.

**Pull request closed without merging**